### ISSUE TITLE
Module not rerunable with ipam

### ISSUE THREAD

**Issue by HanwhaARudolph** [aws-ia/terraform-aws-vpc#65]
When I call the modules to create the VPC and pass it IPAM and rerun the terraform, because the CIDR assigned by ipam changes every time you run the terraform. This appears to be caused by the data call:
```
data "aws_vpc_ipam_preview_next_cidr" "main" {
  count = var.vpc_ipv4_ipam_pool_id == null ? 0 : 1

  ipam_pool_id   = var.vpc_ipv4_ipam_pool_id
  netmask_length = var.vpc_ipv4_netmask_length
}
```
If you rerun the module after creating a VPC, it gets a new CIDR and tries to destroy the VPC and all resources because the VPC cidr "changed". 

## example (run the ipam first and then uncomment the VPC module call and run it twice more)
```
variable "network_map" {
  description = "map of network modules, check variable details on modules/network"
  type        = map(any)
  default = {
    "ci" = {
      suffix = "ci_vpc"
      ipam_region = "us-west-2/non_prod"
      netmask_length = 22
    },
  }
}

module "create_ipam" {
  source   = "github.com/aws-ia/terraform-aws-ipam"
  
  top_cidr = ["10.0.0.0/8"]
  top_name = "Global ipam"

  pool_configurations = {
    us-west-2 = {
      description = "2nd level, locale us-west-2 pool"
      cidr        = ["10.0.0.0/14"]
      locale = "us-west-2"

      sub_pools = {
        non_prod = {
          name                 = "non_prod_ipam"
          cidr                 = ["10.0.0.0/16"]
        }

        prod = {
          name                 = "prod_ipam"
          cidr                 = ["10.1.0.0/16"]
        }
      }
    },
  }
}

module "vpc" {
  source  = "aws-ia/vpc/aws"
  version = ">= 1.4.1"

  for_each = var.network_map

  name     = each.value.suffix
  az_count = 2

  vpc_ipv4_ipam_pool_id   = module.create_ipam.pools_level_2[each.value.ipam_region].id
  vpc_ipv4_netmask_length = each.value.netmask_length

  subnets = {
    public = {
      netmask                   = 24
      nat_gateway_configuration = "all_azs"
    }
    private = {
      netmask      = 24
      route_to_nat = true
    }
  }
}
```

## additional problematic behavior
Because it uses the data command, if a ipam pool is "full" even the destroy command will fail because it tries to acquire a new ipam and it throws that there isn't enough room in ipam given the netmask.

The data command also becomes an issue when you try to spool ipam and VPC in the same module (for the multi-account setup I'm building with the ipam and VPC stuff in a central "networking" account). It throws the following error:
```
│ Error: Invalid count argument
│
│   on .terraform/modules/vpc/data.tf line 41, in data "aws_vpc_ipam_preview_next_cidr" "main":
│   41:   count = var.vpc_ipv4_ipam_pool_id == null ? 0 : 1
│
│ The "count" value depends on resource attributes that cannot be determined until apply, so Terraform cannot predict how many instances will be created. To work around this, use the -target argument to first apply only the resources that the count depends on.
```

**Comment by drewmullen** [comment]
bah! i was supposed to ignore that change. I'll get a fix pushed for this today. Thank you for opening this issue

**Comment by drewmullen** [comment]
Seems like there are 3 issues here:

1. 2nd apply is not working with ipam arguments (confirmed): This is definitely a problem and Im looking into how to fix that

2. If a pool cannot query for another CIDR the data source fails: This is interesting and may require an update to the provider to account. Can you please post the error you get when you run into the full event

3. Cannot create an ipam in the same root module. This is not so much of a bug but i agree we could potentially fix this race condition by referencing a non-computed value; for example we could use a bool `ipam = bool` and that would unblock the dependency issue

One thing to note. I do not believe you would be able to use `for_each` to loop over creating these resources if referencing ipam. The reason is because the query for a CIDR is done with the data source before the resource graph is built so each VPC will get the same CIDR block. There is, sadly, no easy way around this :/ The vpc resource can receive the next computed CIDR by specifying those arguments, however, then we cannot compute the subnets because theyre an attribute of the VPC resource itself (race condition)

**Comment by HanwhaARudolph** [comment]
Yea I have been racking my brain to try to come up with a solution for this type of thing so I am debating using some sort of cidr calculator to parse the values first but since terraform is single pass I'm struggling to find a good solution. If I come up with something I'll add it as a comment here. I will also try to reproduce the error I saw next week (got some issues I'm working on right now).

**Comment by drewmullen** [comment]
Yeah im struggling to fix the 2nd apply issue (and im sad i didnt catch this before now 😭 ). I can get the resources to work and require no changes but the outputs still change because the calculator still fires

**Comment by drewmullen** [comment]
Alright, i got a fix coming regarding the 2nd apply issue  but its going to be part of v2 which will have some **breaking changes** because im changing some inputs and outputs

I might release as alpha next week, if i have time to finish it... would love your input prior to raising it to GA if youre interested

**Comment by HanwhaARudolph** [comment]
> Alright, i got a fix coming regarding the 2nd apply issue but its going to be part of v2 which will have some **breaking changes** because im changing some inputs and outputs
> 
> I might release as alpha next week, if i have time to finish it... would love your input prior to raising it to GA if youre interested

100%
I'd definitely give it a whirl on Monday. Just let me know the branch name or tag me on the PR.

**Comment by drewmullen** [comment]
I've rethought my implementation a bit regarding how to solve this issue. I proposed to hashi we add a modifier to the data source that prevents preview always grabbing a new CIDR. I dont think that behavior is even useful.

This however is bending the notion of a data source so I'm not certain how theyll respond.

**Comment by HanwhaARudolph** [comment]
> I've rethought my implementation a bit regarding how to solve this issue. I proposed to hashi we add a modifier to the data source that prevents preview always grabbing a new CIDR. I dont think that behavior is even useful.
> 
> This however is bending the notion of a data source so I'm not certain how theyll respond.

Yea I won't hold my breath on it. I am going to work on a cidr calculator for working with foreach loops.

**Comment by drewmullen** [comment]
I tried implementing on my own locally but we're limited by what terraform core allows: https://github.com/hashicorp/terraform/issues/31549

I'm going back to my original idea without `run_once` and just a bunch of greedy `ignore_changes` :/ 

Hopefully will have that out for testing in a day or 2

**Comment by drewmullen** [comment]
issue closed by default. i got a couple more PRs coming before we cut a 2.0 release but, if youre able, would love your feedback on whats currently in main from #67

**Comment by HanwhaARudolph** [comment]
Works great now for re-runability but if you want to move something it does fail (just changed the region and ipam region and got the following error):
```
│ Error: Error previewing next cidr from IPAM pool (ipam-pool-XXXXXXXXXX): InvalidParameterValue: The pool ipam-pool-XXXXXXXXXXXXX is not in the current region us-west-2.
│       status code: 400, request id: e83e2f8a-f81a-44a8-bccf-f59ca5936b21
│
│   with module.ci_vpc.data.aws_vpc_ipam_preview_next_cidr.main[0],
│   on .terraform/modules/ci_vpc/data.tf line 53, in data "aws_vpc_ipam_preview_next_cidr" "main":
│   53: data "aws_vpc_ipam_preview_next_cidr" "main" {
```

but that is out of scope from the issue I raised and something that I think will happen no matter what unless you target destroy first and then respool.

**Comment by drewmullen** [comment]
I was also able to refactor a bit which should help with refresh times and allowed us to use the RESOURCE instead of the DATA SOURCE. ik that was causing an error for you when the pool ran out of cidrs. I had never considered that situation!

it also allows us to remove all the greedy `ignore_changes`

https://github.com/aws-ia/terraform-aws-vpc/pull/69

**Comment by drewmullen** [comment]
v2.0 release was cut today so you should be able to use ipam as expected. we eliminated 2 of your problems:

1. you can now perform second apply with no issue
2. we do not use the data source and therefor do not query for multiple cidrs, you should not see the error when your pool is out of ranges
